plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/subnets"
+	"github.com/gophercloud/gophercloud/openstack/orchestration/v1/stacks"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"log"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+const (
+	// plugin is used to identify ourselves in log output
+	plugin = "openstack"
+)
+
+// serviceType is an OpenStack service type
+type serviceType string
+
+const (
+	volumeV2Service serviceType = "volumev2"
+	Orchestration   serviceType = "orchestration"
+)
+
+// tagMap maps a tag name to value.
+type tagMap map[string]string
+
+// fieldMap maps a field to an arbitrary data type.
+type fieldMap map[string]interface{}
+
+// serverDiag is a map of server diagnostics fields
+type serverDiag map[string]interface{}
+
+// serviceMap maps service id to a Service struct.
+type serviceMap map[string]services.Service
+
+// stackMap maps stack id to a Stack struct.
+type stackMap map[string]stacks.ListedStack
+
+// nova_serviceMap maps nova_service id to a nova Service struct.
+type nova_serviceMap map[string]nova_services.Service
+
+// ContainsService indicates whether a particular service is enabled
+func (s serviceMap) ContainsService(t serviceType) bool {
+	for _, service := range s {
+		if service.Type == string(t) {
+			return true
+		}
+	}
+	return false
+}
+
+// projectMap maps a project id to a Project struct.
+type projectMap map[string]projects.Project
+
+// hypervisorMap maps a hypervisor id a Hypervisor struct.
+type hypervisorMap map[string]hypervisors.Hypervisor
+
+// serverMap maps a server id to a Server struct.
+type serverMap map[string]servers.Server
+
+// flavorMap maps a flavor id to a Flavor struct.
+type flavorMap map[string]flavors.Flavor
+
+// subnetMap maps a subnet id to a Subnet struct.
+type subnetMap map[string]subnets.Subnet
+
+// portMap maps a port id to a Port struct.
+type portMap map[string]ports.Port
+
+// networkMap maps a network id to a Network struct.
+type networkMap map[string]networks.Network
+
+// aggregateMap maps a aggregate id to a Aggregate struct.
+type aggregateMap map[int]aggregates.Aggregate
+
+// agentMap maps a agent id to a agents struct.
+type agentMap map[string]agents.Agent
+
+// volume is a structure used to unmarshal raw JSON from the API into.
+type volume struct {
+	volumes.Volume
+	volumetenants.VolumeTenantExt
+}
+
+// volumeMap maps a volume id to a volume struct.
+type volumeMap map[string]volume
+
+// storagePoolMap maps a storage pool name to a StoragePool struct.
+type storagePoolMap map[string]schedulerstats.StoragePool
+
+// OpenStack is the main structure associated with a collection instance.
+type OpenStack struct {
+	// Configuration variables
+	IdentityEndpoint   string
+	Domain             string
+	Project            string
+	Username           string
+	Password           string
+	EnabledServices    []string
+	ServerDiagnotics   bool
+	InsecureSkipVerify bool
```

Comment:
You should probably include `tls.ClientConfig`. Take a look at e.g. the prometheus input plugin on how to use it.